kyo-prelude/shared/src/main/scala/kyo2/kernel/Effect.scala

```diff
@@ -165,6 +162,51 @@ object Effect:
             handle3Loop(v, Context.empty)
         end apply
 
+        private[kyo2] inline def partial[I1[_], O1[_], E1 <: Effect[I1, O1], I2[_], O2[_], E2 <: Effect[I2, O2], I3[_], O3[_], E3 <: Effect[
+            I3,
+            O3
+        ], A, S, S2](
+            inline tag1: Tag[E1],
+            inline tag2: Tag[E2],
+            inline tag3: Tag[E3],
+            v: A < (E1 & E2 & E3 & S),
+            context: Context
+        )(
+            inline stop: => Boolean,
+            inline handle1: Safepoint ?=> [C] => (I1[C], O1[C] => A < (E1 & E2 & E3 & S & S2)) => A < (E1 & E2 & E3 & S & S2),
+            inline handle2: Safepoint ?=> [C] => (I2[C], O2[C] => A < (E1 & E2 & E3 & S & S2)) => A < (E1 & E2 & E3 & S & S2),
+            inline handle3: Safepoint ?=> [C] => (I3[C], O3[C] => A < (E1 & E2 & E3 & S & S2)) => A < (E1 & E2 & E3 & S & S2)
+        )(using inline _frame: Frame, safepoint: Safepoint): A < (E1 & E2 & E3 & S & S2) =
+            def partialLoop(v: A < (E1 & E2 & E3 & S & S2), context: Context)(using safepoint: Safepoint): A < (E1 & E2 & E3 & S & S2) =
+                if stop then v
+                else
+                    v match
+                        case <(kyo: KyoSuspend[?, ?, ?, ?, ?, ?]) =>
+                            if kyo.tag =:= Tag[Defer] then
+                                val k = kyo.asInstanceOf[KyoSuspend[Const[Unit], Const[Unit], Defer, Any, A, Any]]
+                                partialLoop(k((), context), context)
+                            else
+                                safepoint.pushFrame(kyo.frame)
+                                if tag1 =:= kyo.tag then
+                                    val k = kyo.asInstanceOf[KyoSuspend[I1, O1, E1, Any, A, E1 & E2 & E3 & S & S2]]
```

Comment:
Perhaps helpful to make a `type Suspend[I, O, E] = KyoSuspend[I, O, E, Any, A, E1 & E2 & E3 & S & S2]` on line 189?

Reply:
nice cleanup! I also changed the `Defer` branch to use it.